src/library/scala/collection/immutable/Stream.scala

```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
   private[immutable] def collectedTail[A, B, That](head: B, stream: Stream[A], pf: PartialFunction[A, B], bf: CanBuildFrom[Stream[A], B, That]) = {
     cons(head, stream.tail.collect(pf)(bf).asInstanceOf[Stream[B]])
   }
-}
-
 
+  /** An implementation of `FilterMonadic` allowing GC of the `Stream` as it is processed.
+    *
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```

Comment:
`withFilter` was introduced to reduce intermediate collection creation in for expressions, but the method is available for anyone to call in regular code to.

![image](https://cloud.githubusercontent.com/assets/65551/6054253/a0ebb5f8-ad36-11e4-90a1-d06d8bcb184a.png)

I think you are correct in your assessment that we can't have an immutable StreamWithFilter *and* avoid retaining the head reference.